File: bindings/java-jni/src/test/java/org/maplibre/nativejni/internal/callback/ResourceTransformStateTest.java
package org.maplibre.nativejni.internal.callback;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import org.bytedeco.javacpp.BytePointer;
import org.junit.jupiter.api.Test;
import org.maplibre.nativejni.internal.javacpp.MaplibreNativeC;
import org.maplibre.nativejni.resource.ResourceKind;
import org.maplibre.nativejni.resource.ResourceTransformRequest;

// Support invariant for BND-121, BND-123, BND-140, and BND-141: these tests exercise
// Java JNI's callback stub directly to force host failures and in-flight close races.
class ResourceTransformStateTest {
  @Test
  void bnd123CloseWaitsForInflightCallbackBeforeClosingStub() throws Exception {
    var entered = new CountDownLatch(1);
    var release = new CountDownLatch(1);
    var closed = new AtomicBoolean();
    var state =
        new ResourceTransformState(
            request -> {
              entered.countDown();
              assertTrue(await(release));
              return java.util.Optional.of(request.url() + "?rewritten=true");
            });

    try (var url = new BytePointer("https://example.com/style.json");
        var response = new MaplibreNativeC.mln_resource_transform_response()) {
      response.size(response.sizeof());
      response.url(null);

      var callbackThread =
          new Thread(
              () ->
                  state
                      .transform()
                      .callback()
                      .call(null, ResourceKind.STYLE.nativeValue(), url, response));
      callbackThread.start();
      assertTrue(await(entered));

      var closeThread =
          new Thread(
              () -> {
                state.close();
                closed.set(true);
              });
      closeThread.start();

      Thread.sleep(25);
      assertFalse(closed.get());
      release.countDown();
      callbackThread.join();
      closeThread.join();

      assertTrue(closed.get());
    }
  }

  @Test
  void bnd123CloseFromCurrentCallbackFailsWithoutDeadlock() {
    var attempted = new AtomicBoolean();
    var state = new AtomicReference<ResourceTransformState>();
    state.set(
        new ResourceTransformState(
            request -> {
              attempted.set(true);
              state.get().close();
              return java.util.Optional.empty();
            }));

    try (var url = new BytePointer("https://example.com/style.json");
        var response = new MaplibreNativeC.mln_resource_transform_response()) {
      response.size(response.sizeof());
      response.url(null);

      var status =
          state
              .get()
              .transform()
              .callback()
              .call(null, ResourceKind.STYLE.nativeValue(), url, response);

      assertEquals(MaplibreNativeC.MLN_STATUS_NATIVE_ERROR, status);
      assertTrue(attempted.get());
      assertFalse(state.get().isClosed());
    } finally {
      state.get().close();
    }
  }

  @Test
  void bnd121InvalidRewriteUrlFallsBackToPassThrough() {
    try (var state = new ResourceTransformState(request -> java.util.Optional.of("bad\0url"));
        var url = new BytePointer("https://example.com/style.json");
        var response = new MaplibreNativeC.mln_resource_transform_response()) {
      response.size(response.sizeof());
      response.url(null);

      var status =
          state.transform().callback().call(null, ResourceKind.STYLE.nativeValue(), url, response);

      assertEquals(MaplibreNativeC.MLN_STATUS_OK, status);
      assertTrue(response.url() == null || response.url().isNull());
    }
  }

  @Test
  void bnd121CallbackExceptionReturnsNativeErrorAndNoRewrite() {
    try (var state =
            new ResourceTransformState(
                request -> {
                  throw new IllegalStateException("boom");
                });
        var url = new BytePointer("https://example.com/style.json");
        var response = new MaplibreNativeC.mln_resource_transform_response()) {
      response.size(response.sizeof());
      response.url(null);

      var status =
          state.transform().callback().call(null, ResourceKind.STYLE.nativeValue(), url, response);

      assertEquals(MaplibreNativeC.MLN_STATUS_NATIVE_ERROR, status);
      assertTrue(response.url() == null || response.url().isNull());
    }
  }

  @Test
  void bnd141RequestCopiesBorrowedUrlAndPreservesRawKind() {
    var captured = new AtomicReference<ResourceTransformRequest>();
    try (var state =
            new ResourceTransformState(
                request -> {
                  captured.set(request);
                  return java.util.Optional.empty();
                });
        var url = new BytePointer("https://example.com/source.json");
        var response = new MaplibreNativeC.mln_resource_transform_response()) {
      response.size(response.sizeof());
      response.url(null);

      var status = state.transform().callback().call(null, 0x7fff, url, response);

      assertEquals(MaplibreNativeC.MLN_STATUS_OK, status);
    }

    assertEquals(0x7fff, captured.get().kind().rawValue());
    assertEquals("https://example.com/source.json", captured.get().url());
  }

  private static boolean await(CountDownLatch latch) {
    try {
      return latch.await(5, TimeUnit.SECONDS);
    } catch (InterruptedException exception) {
      Thread.currentThread().interrupt();
      return false;
    }
  }
}

File: bindings/kotlin/src/nativeTest/kotlin/org/maplibre/nativeffi/resource/ResourceTransformStateTest.kt
package org.maplibre.nativeffi.resource

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.alloc
import kotlinx.cinterop.cstr
import kotlinx.cinterop.memScoped
import kotlinx.cinterop.ptr
import org.maplibre.nativeffi.error.MaplibreStatus
import org.maplibre.nativeffi.internal.c.mln_resource_transform_response
import org.maplibre.nativeffi.internal.callback.ResourceTransformState

@OptIn(ExperimentalForeignApi::class)
class ResourceTransformStateTest : org.maplibre.nativeffi.NativeTestBase() {
  @Test
  fun transformCallbackCopiesRequestAndPreservesUnknownKindRawValue() {
    var copiedRequest: ResourceTransformRequest? = null
    val state =
      ResourceTransformState(
        ResourceTransformCallback { request ->
          copiedRequest = request
          null
        }
      )
    try {
      memScoped {
        val out = alloc<mln_resource_transform_response>()
        val status =
          state.invoke(999U, "https://example.com/style.json".cstr.getPointer(this), out.ptr)
        assertEquals(MaplibreStatus.OK.nativeCode, status)
        assertNull(out.url)
      }
      assertEquals(ResourceKind(999), copiedRequest?.kind)
      assertEquals(999, copiedRequest?.kind?.nativeValue)
      assertEquals("https://example.com/style.json", copiedRequest?.url)
    } finally {
      state.close()
    }
  }

  @Test
  fun embeddedNulReplacementReturnsInvalidArgument() {
    val state =
      ResourceTransformState(
        ResourceTransformCallback { "https://cdn.example.com/\u0000truncated" }
      )
    try {
      memScoped {
        val out = alloc<mln_resource_transform_response>()
        assertEquals(MaplibreStatus.INVALID_ARGUMENT.nativeCode, state.invoke(1U, null, out.ptr))
      }
    } finally {
      state.close()
    }
  }

  @Test
  fun nullEmptyAndThrowingTransformsProduceNoReplacement() {
    memScoped {
      val nullState = ResourceTransformState(ResourceTransformCallback { null })
      val nullOut = alloc<mln_resource_transform_response>()
      assertEquals(MaplibreStatus.OK.nativeCode, nullState.invoke(3U, null, nullOut.ptr))
      assertNull(nullOut.url)
      nullState.close()

      // BND-121: host-language failures do not escape the C callback boundary.
      val throwingState =
        ResourceTransformState(
          ResourceTransformCallback { throw IllegalStateException("contained") }
        )
      val throwingOut = alloc<mln_resource_transform_response>()
      assertEquals(
        MaplibreStatus.NATIVE_ERROR.nativeCode,
        throwingState.invoke(3U, null, throwingOut.ptr),
      )
      throwingState.close()
    }
  }

  @Test
  fun closeDuringTransformCallbackCompletesAfterCallbackAndSuppressesLaterUpcalls() {
    var calls = 0
    lateinit var state: ResourceTransformState
    state =
      ResourceTransformState(
        ResourceTransformCallback {
          calls += 1
          state.close()
          assertFalse(state.isClosedForTesting())
          null
        }
      )
    memScoped {
      val out = alloc<mln_resource_transform_response>()

      assertEquals(MaplibreStatus.OK.nativeCode, state.invoke(1U, null, out.ptr))
      assertTrue(state.isClosedForTesting())
      assertEquals(MaplibreStatus.INVALID_ARGUMENT.nativeCode, state.invoke(1U, null, out.ptr))
      assertEquals(1, calls)
      state.close()
    }
  }
}
